daemon/src/collab_settlement_taker.rs

```diff
@@ -1,78 +1,106 @@
 use crate::address_map::ActorName;
 use crate::address_map::Stopping;
+use crate::cfd_actors::load_cfd;
 use crate::connection;
-use crate::model::cfd::Cfd;
+use crate::model::cfd;
+use crate::model::cfd::CfdEvent;
 use crate::model::cfd::CollaborativeSettlement;
 use crate::model::cfd::Completed;
+use crate::model::cfd::OrderId;
 use crate::model::cfd::SettlementKind;
 use crate::model::cfd::SettlementProposal;
 use crate::model::Price;
+use crate::process_manager;
 use crate::projection;
 use crate::send_async_safe::SendAsyncSafe;
 use crate::wire;
 use anyhow::Result;
 use async_trait::async_trait;
-use xtra::prelude::MessageChannel;
 use xtra_productivity::xtra_productivity;
 
 pub struct Actor {
-    cfd: Cfd,
-    projection: xtra::Address<projection::Actor>,
-    on_completed: Box<dyn MessageChannel<Completed<CollaborativeSettlement>>>,
+    proposal: Option<SettlementProposal>,
+    order_id: OrderId,
+    current_price: Price,
+    n_payouts: usize,
     connection: xtra::Address<connection::Actor>,
-    proposal: SettlementProposal,
+    process_manager: xtra::Address<process_manager::Actor>,
+    projection: xtra::Address<projection::Actor>,
+    db: sqlx::SqlitePool,
 }
 
 impl Actor {
     pub fn new(
-        cfd: Cfd,
-        projection: xtra::Address<projection::Actor>,
-        on_completed: impl MessageChannel<Completed<CollaborativeSettlement>> + 'static,
+        order_id: OrderId,
         current_price: Price,
-        connection: xtra::Address<connection::Actor>,
         n_payouts: usize,
-    ) -> Result<Self> {
-        let proposal = cfd.start_collaborative_settlement_taker(current_price, n_payouts)?;
-
-        Ok(Self {
-            cfd,
-            projection,
-            on_completed: Box::new(on_completed),
+        connection: xtra::Address<connection::Actor>,
+        process_manager: xtra::Address<process_manager::Actor>,
+        projection: xtra::Address<projection::Actor>,
+        db: sqlx::SqlitePool,
+    ) -> Self {
+        Self {
+            proposal: None,
+            order_id,
+            n_payouts,
+            current_price,
             connection,
-            proposal,
-        })
+            process_manager,
+            projection,
+            db,
+        }
     }
 
     async fn propose(&mut self, this: xtra::Address<Self>) -> Result<()> {
+        let mut conn = self.db.acquire().await?;
+        let cfd = load_cfd(self.order_id, &mut conn).await?;
+
+        let event = cfd.propose_collaborative_settlement(self.current_price, self.n_payouts)?;
+        let proposal = if let cfd::Event {
+            event: CfdEvent::CollaborativeSettlementProposed { ref proposal },
+            ..
+        } = event
+        {
+            proposal
+        } else {
+            unreachable!()
+        };
```

Comment:
Damn that is ugly haha! I didn't think of that ...

Reply:
that was the reason I pushed back on similar comment on my review :P returning a tuple would probably be nicer after all, even if we store it both in the event for future use and pass back here!

Reply:
I also tried to avoid this by sending the proposal to the maker in `process_manager::Actor`, but that doesn't quite work because we'd need an address to the `collab_settlement_taker::Actor` which cannot be serialized as part of the event (and also doesn't belong there).